ipq40xx: wpj428 cpximg and reboot

yogo1212 opened the pull request:
hi :-)

see commit description for what the cpximg is used for.
there's also another variant `6A03` around but i haven't got one and can't check which parameters it uses.

the reboot-issue discussed in https://github.com/openwrt/openwrt/pull/787 is fixed like with the unielec-u4019 or like compex https://github.com/compex-systems/lede/blob/openwrt-18.06/target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4028-wpj428.dts#L32.
the workaround with installing another version of u-boot doesn't work anymore since the file is no longer available.

'nudging' the detection isn't enough.
i've also got a patch (based on the compex repo) for sending SPI commands on shutdown but it appears to be working without that ugly thing.
EDIT: 4 byte opcodes have been added in the meantime.

gpio52/53 are EMDC/EMDIO.

adrianschmutzler commented:
Your commits have formal issues. Please ensure proper wrapping of commit message and add a Signed-off.

yogo1212 commented:
@adrianschmutzler done.
sorry for the extra step

adrianschmutzler commented:
Your "From:" includes quotes where your Signed-off doesn't:
```
...
From: "Leon M. George" <leon@georgemail.eu>
...
Signed-off-by: Leon M. George <leon@georgemail.eu>
```

Easiest is to use "git commit --amend --reset-author" which changes from to the current Signed-off.

Despite, you need to add a commit message to your second commit.

yogo1212 committed:
ipq40xx: build wpj428 cpximg

Generate a cpximg that is compatible with the cpximg loader in Compex' u-boot.
The cpximg loader can be started either by holding the reset button during power
up or by entering the u-boot prompt and entering 'cpximg'.
Once it's running, a TFTP-server under 192.168.1.1 will accept an image
appropriate for the board revision that is etched on the board (e.g. 6A04).

The image can be pushed using tftp:
	tftp -v -m binary 192.168.1.1 -c put openwrt-ipq40xx-generic-compex_wpj428-squashfs-cpximg-6A04.bin

cpximg files can also be used with the sysupgrade utility in stock images.
(add SSH key in luci for root access)

Signed-off-by: Leon M. George <leon@georgemail.eu>

yogo1212 commented:
the quotes in "From" are unimpressed by `--reset-author`.
no idea why that happens (using `--signoff`) but if i can't remove the quotes in "From" i'll add them in "Signed-Off"

adrianschmutzler commented:
Well, use
git commit --amend --author "Leon M. George <leon@georgemail.eu>"

The Signed-off can be corrected manually.

yogo1212 commented:
thanks. i'll do that

yogo1212 commented:
@adrianschmutzler 

i've run `git commit --amend --author "Leon M. George <leon@georgemail.eu>"` on both commits but still
```
From dd67b86765dd6905cc1926cdc929078b0ffce76a Mon Sep 17 00:00:00 2001
From: "Leon M. George" <leon@georgemail.eu>
```

my .gitconfig is
```
[user]
  email = leon@georgemail.eu
  name = Leon M. George
```

is there a way we can avoid having to deal with this?

ynezz commented:
@yogo1212 `--reset-author` should do the trick

yogo1212 commented:
```
13:45:05 leon@brot:~/openwrt£ git format-patch HEAD~1 ; head -n2 0001* 
0001-ipq40xx-fix-reboot-for-compex-wpj428.patch
From ee62f6852a21acd932abecd94ea387c7d92af5ac Mon Sep 17 00:00:00 2001
From: "Leon M. George" <leon@georgemail.eu>
13:45:27 leon@brot:~/openwrt£ git commit --amend --reset-author
[wpj428_reboot 69cebda9ca] ipq40xx: fix reboot for compex wpj428
 1 file changed, 17 insertions(+)
13:45:47 leon@brot:~/openwrt£ git format-patch HEAD~1 ; head -n2 0001* 
0001-ipq40xx-fix-reboot-for-compex-wpj428.patch
From 69cebda9ca028c385317674b98978765a5313623 Mon Sep 17 00:00:00 2001
From: "Leon M. George" <leon@georgemail.eu>
13:46:49 leon@brot:~/openwrt£ git config user.name
Leon M. George
```

adrianschmutzler commented:
Strange, I don't see the reason for that. Maybe it's the dot after the middle initial? Try to remove that, and if that's still not working, try to remove the middle initial entirely (at least for testing).

yogo1212 commented:
If i change my name to `Leon George`, there are no quotes.

```
13:52:53 leon@brot:~/openwrt£ git config user.name Leon\ George
13:54:57 leon@brot:~/openwrt£ git commit --amend --reset-author
[wpj428_reboot 19eab9456c] ipq40xx: fix reboot for compex wpj428
 1 file changed, 17 insertions(+)
13:55:02 leon@brot:/openwrt£ git format-patch HEAD~1 ; head -n2 0001* 
0001-ipq40xx-fix-reboot-for-compex-wpj428.patch
From 19eab9456c2a270e2a8906f9ae5dacefd145d89a Mon Sep 17 00:00:00 2001
From: Leon George <leon@georgemail.eu>
```

adrianschmutzler commented:
And for "Leon M George"?

yogo1212 commented:
```
13:57:42 leon@brot:~/openwrt£ git config user.name Leon\ M\ George
14:09:00 leon@brot:~/openwrt£ git commit --amend --reset-author --no-edit
[wpj428_reboot 9dcbe79acd] ipq40xx: fix reboot for compex wpj428
 1 file changed, 17 insertions(+)
14:09:03 leon@brot:~/openwrt£ git format-patch HEAD~1 ; head -n2 0001* 
0001-ipq40xx-fix-reboot-for-compex-wpj428.patch
From 9dcbe79acd8e59d71b2230a297b86439e71c15a9 Mon Sep 17 00:00:00 2001
From: Leon M George <leon@georgemail.eu>
```

yogo1212 committed:
ipq40xx: fix reboot for compex wpj428

Reset the MX25l25635F NOR flash on boot using a dedicated GPIO.

Signed-off-by: Leon M. George <leon@georgemail.eu>

adrianschmutzler commented:
Okay, so looks like the quotes are protecting the dots so it's not parsed as mail address or domain or whatever.

I've looked in the repo, and to me it looks like patches have been accepted with quotes already:
https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=f974f8213b94578581b35e6b3f8fb1fd5a35f753

In the git.openwrt.org view, the do not look malformatted by any means, so I think just keep the version with dot and quotes if you like, or remove the dot if you don't like the quotes ...

yogo1212 commented:
am i the only one using an abbreviated middle name? oO
other complicated names use escape-characters `From: =?UTF-8?q?Old=C5=99ich=20Jedli=C4=8Dka?= <oldium.pro@gmail.com>` (e0ce80d42).
i would argue that it isn't sensible to have matching formats in "Signed-off" and "From".

quotes in the mail-header don't bother me; afaics, they're generated and the commit created from the patch doesn't have quotes - at least for me.
the signed-off line is part of the comment which is supposed to be for humans to read, so i wouldn't put any quotes there.

the latest commits are formatted the way git created them by default which seems appropriate.

yogo1212 commented:
@adrianschmutzler in 6abce4d5e, phy-reset-gpio was removed in favor of `reset-gpios`.
the commits here have been updated accordingly.

yogo1212 commented:
getting this error:
```
[    0.681921] libphy: ipq40xx_mdio: probed
[    0.684277] mdio_bus 90000.mdio: MDIO device at address 0 is missing.
[    0.688200] mdio_bus 90000.mdio: MDIO device at address 1 is missing.
[    0.694560] mdio_bus 90000.mdio: MDIO device at address 2 is missing.
[    0.700961] mdio_bus 90000.mdio: MDIO device at address 3 is missing.
[    0.707405] mdio_bus 90000.mdio: MDIO device at address 4 is missing.
[    0.713872] ar40xx c000000.ess-switch: Probe failed - Missing PHYs!
[    0.720245] libphy: Fixed MDIO Bus: probed
[    0.730040] libphy: PHY 90000.mdio:04 not found
[    0.803925] ------------[ cut here ]------------
[    0.803973] WARNING: CPU: 1 PID: 1 at mm/vmalloc.c:1515 release_nodes+0x1a0/0x1c4
[    0.807615] Trying to vfree() nonexistent vm area ((ptrval))
[    0.815026] Modules linked in:
[    0.820721] CPU: 1 PID: 1 Comm: swapper/0 Not tainted 4.19.97 #0
[    0.823584] Hardware name: Generic DT based system
[    0.829767] [<c030e310>] (unwind_backtrace) from [<c030ab70>] (show_stack+0x10/0x14)
[    0.834365] [<c030ab70>] (show_stack) from [<c0785260>] (dump_stack+0x94/0xa8)
[    0.842264] [<c0785260>] (dump_stack) from [<c031bb40>] (__warn+0xd4/0xf0)
[    0.849290] [<c031bb40>] (__warn) from [<c031b7e8>] (warn_slowpath_fmt+0x38/0x48)
[    0.856152] [<c031b7e8>] (warn_slowpath_fmt) from [<c05aaafc>] (release_nodes+0x1a0/0x1c4)
[    0.863702] [<c05aaafc>] (release_nodes) from [<c05a75e4>] (really_probe+0x1d0/0x2b4)
[    0.871859] [<c05a75e4>] (really_probe) from [<c05a7988>] (__driver_attach+0xa0/0xd4)
[    0.879762] [<c05a7988>] (__driver_attach) from [<c05a5af4>] (bus_for_each_dev+0x48/0x90)
[    0.887573] [<c05a5af4>] (bus_for_each_dev) from [<c05a6af0>] (bus_add_driver+0xec/0x1fc)
[    0.895730] [<c05a6af0>] (bus_add_driver) from [<c05a8064>] (driver_register+0xbc/0xf8)
[    0.903891] [<c05a8064>] (driver_register) from [<c030262c>] (do_one_initcall+0x7c/0x184)
[    0.911707] [<c030262c>] (do_one_initcall) from [<c0a00ef8>] (kernel_init_freeable+0x1c0/0x25c)
[    0.920039] [<c0a00ef8>] (kernel_init_freeable) from [<c07975cc>] (kernel_init+0x8/0x110)
[    0.928543] [<c07975cc>] (kernel_init) from [<c03010e8>] (ret_from_fork+0x14/0x2c)
[    0.936869] Exception stack(0xcf839fb0 to 0xcf839ff8)
[    0.944333] 9fa0:                                     00000000 00000000 00000000 00000000
[    0.949462] 9fc0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[    0.957621] 9fe0: 00000000 00000000 00000000 00000000 00000013 00000000
[    0.965812] ---[ end trace edfa8f699df36cce ]---
[    0.972215] ess_edma: probe of c080000.edma failed with error -5
```
with or without the patch, there's no ethernet for me :-(

yogo1212 commented:
0 != ACTIVE_LOW :neutral_face: :palm_tree:

adrianschmutzler commented:
Indeed.

yogo1212 commented:
having compensated my copy'n'paste-negligence, i can proudly proclaim that the fix for rebooting the WPJ428 no longer breaks ethernet.

T-X commented:
Overall looks good to me.

Just one thing: A quick test on a python console gives me a float: ">>> math.ceil(1234/64) \ 19.0". And a ">>> print(hex(19.0))" returns an error for me. Here on my Debian Sid python symlinks to python2.7. With python3 the math.ceil returns a plain integer to me with no decimal points and therefore does not crash.

Also the python documentation seems to indicate that in 2.7 it's always a float: https://docs.python.org/2.7/library/math.html#math.ceil

Should this maybe be changed to use python3 explicitly? (Or am I missing something?)

adrianschmutzler commented:
OpenWrt requires python 3.5+, but I do not know how this is enforced.

adrianschmutzler commented:
I still would like to get rid of the python dependence at all here anyway.

adrianschmutzler commented:
What you need is:
```
let \
		size="$$(stat -c%s $@)" \
		pad="$(subst k,* 1024,$(word 3, $(1)))" \
		pad="(pad - (size % pad)) % pad" \
		newsize='size + pad'
```

Then you would just append "64k" to mkmylofw, or you could use BLOCKSIZE directly in the build recipe. Despite, it might make sense to put the size calculation into a separate function.

Maybe one could just create some padded-size definition:

```
define padded_size
let \
		size="$$(stat -c%s $(1))" \
		pad="$(subst k,* 1024,$(BLOCKSIZE))" \
		pad="(pad - (size % pad)) % pad" \
		newsize='size + pad'; \
		printf "0x%x" $newsize
endef
```

and then

`-p0x180000:$(call padded_size,$@):al:0x84000000`

(Only partially tested ...)

adrianschmutzler commented:
Please also check out my comments about just using padding instead:
https://github.com/openwrt/openwrt/pull/2265#issuecomment-589273386

adrianschmutzler commented:
I've pushed an updated version here:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/compex2

If that doesn't work out-of-the-box, you may try with additional pad-to:

`IMAGE/cpximg-6a04.bin := append-kernel | append-rootfs | pad-rootfs | pad-to $$$$(BLOCKSIZE) | mkmylofw_32m 0x8A2 3`

Other questions:

> cpximg files can also be used with the sysupgrade utility in stock images.

Does that mean they are effectively "factory" images, and can be flashed directly from stock without the bootloader?

> (add SSH key in luci for root access)

What's that supposed to mean?

> 0x84000000

What's that referring to (in the mkmylofw -p argument)?

adrianschmutzler commented:
Sorry, pushed an outdated version before, now it should be up-to-date.

adrianschmutzler commented:
@yogo1212 mkmylofw tells me for -p:
```
" -p <addr>:<len>[:<flags>[:<param>[:<name>[:<file>]]]]\n"
"                  add partition at <addr>, with size of <len> to the\n"
"                  partition table, ...
```

I wonder whether whether the "size of the partition" should be the size of the partition instead of your custom padding ...

yogo1212 commented:
@T-X @adrianschmutzler  openwrt's PATH causes the python in `staging_dir/host/bin/` to be used.
try `ls -al staging_dir/host/bin/python`. that should be python3.
regardless, the other points you make are reason enough to ditch the python code.

> I wonder whether whether the "size of the partition" should be the size of the partition instead of your custom padding ...

Technically, there are really two 'partitions' - as far as u-boot is concerned. The partitioning paradigm is really something we're ignoring/working against.
Imho, the important aspect is that increasing the partition size on the cpximg-layer effectively doubles the time it takes to flash a board since the whole region of the flash is erased instead of just the portion that would be needed.

The cpximg format is a good tool for a manufacturing context.
For an individual flashing maybe one or two boards, time might not be an issue.
But for manufacturing processes, it is very important.

It appears compex is using mkmylofw as well. They don't use it as a partition size. I suspect they use partitions to install u-boot, caldata, etc. They don't appear to use the second partition either.
In their words:
".. the size is round off to 0x580000 as the size of the sysupgrade.bin is 0x57xxxx .."

> If that doesn't work out-of-the-box, you may try with additional pad-to

The `pad-to` is necessary, else (with only pad-root):
```
bc  <<< "obase=16; $(stat -c%s $@)"
8C01EC
```
And without the two LSBs being 0, cpximg loader doesn't accept the file.

Considering the latest additions to the discussion, i favour your solution:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=blobdiff;f=target/linux/ipq40xx/image/Makefile;h=7ad370a66d708fbbe774f91d12eb2ef093aca004;hp=bbe4df9df40bea350038ecfa52b1cf381eee4142;hb=72a162b7bea0973d7fd7d972b6fecee3447800ff;hpb=2f1cc5c3d5e35d6aa76e794e3d5b4f5856cd38bc

The hex format is not necessary, since mkmylofw uses `strtoul(.., 0)`.
I used it mainly for easier debugging (making sure the LSBs bytes were 0). With the 6 characters in python that luxury seemed reasonable. With the padding conundrum handled outside of that function, there's no need to have it there.

> Does that mean they are effectively "factory" images, and can be flashed directly from stock without the bootloader?
They can be flashed without dealing with u-boot, yes. "Factory"? idk

> [ssh key, luci] What's that supposed to mean?
The stock firmware has luci installed and, afaik, setting the ssh key there is the only way to get access to a root shell.

> [0x84000000] What's that referring to (in the mkmylofw -p argument)?
I suspect it's the RAM address where the block should be stored.

adrianschmutzler commented:
> I strongly suspect it's the RAM address where the block should be temporarily stored. It works with other addresses as well.

I've been digging a little deeper:

https://github.com/openwrt/openwrt/blob/1d47f81581bcd476487e5b29314602167a843b99/tools/firmware-utils/src/myloader.h#L109

/* if this is the active partition, the MyLoader load code to this address */

I'm just asking because the other two (very old) cases of mkmylofw use smaller addresses:
https://github.com/openwrt/openwrt/blob/34939711a51733d6461a0a653776c6de0804fa48/target/linux/ath25/image/Makefile#L29
https://github.com/openwrt/openwrt/blob/99835e09992bb411519fda1f8d16a6d07df7f598/target/linux/ar71xx/image/legacy.mk#L541

adrianschmutzler commented:
^ For ar71xx the param value is the KERNEL_LOADADDR.

yogo1212 commented:
Good stuff!

adrianschmutzler commented:
Okay, I've prepared and build-tested two versions:
With pad-to: https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/compex2
Without pad-to: https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/compex3

I've changed the param to ipq40xx KERNEL_LOADADDR in both cases, though I'm not sure whether this is really an improvement or just doesn't matter at all.

Please run-test and tell me if you have a preference for one of the versions.

yogo1212 commented:
the images work. maybe the load address doesn't matter?

arguing just from the size-point-of-view:
```
13:17:33 leon@brot:/media/openwrt£ la compex*
-rw-r--r-- 1 leon users 9568744 Mar  6 11:59 compex2.bin
-rw-r--r-- 1 leon users 9503212 Mar  6 12:52 compex3.bin
13:17:36 leon@brot:/media/openwrt£ echo $((9568744 - 9503212))
65532
```
compex2 will need 127 more tftp block transfers (65532 / 512 = 127, default tftp block size). because pad-rootfs always leaves the file size just above block boundary, this overhead is static.
the solution for compex3 is significantly less elegant, though...

not having to make a decision made having a strong opinion easier :wink: 
hmm...

i'll still go with my argument - a big point of providing the cpximg file is enabling a scalable installation process for openwrt.
tftp can only transfer blocks individually and sequentially. additional block not only adds a fixed round-trip but a potential re-transmission as well, thus reducing the probability of a successfull file transaction within a given time frame.

adrianschmutzler commented:
I've just merged compex3 variant.

Since this PR is crowded with formal issues and mostly the cpximg discussion, and since we have kernel 5.4 now, I'd (as an exception) recommend to resubmit the reboot fix (for kernel 4.19 and 5.4 if applicable) either as a new PR or via the mailing list. Otherwise, it would just get lost here.

Consequently, I close this PR now. Thanks for keeping up.

yogo1212 commented:
@adrianschmutzler fair enough, thank you :-)
